blogposts/2021/postgres-version-update.md

```diff
@@ -0,0 +1,139 @@
+
+Each migration is written in a generic way that allows it to be controlled by a global migration runner. This runner is responsible for controlling the throughput of the migration so that it does not require disproportionate memory or compute resources in the database or in the application.
+
+Each migration is required to report its own progress (0-100%) to aid the runner in determining the set of migrations which need to run (and in which direction they must run in the presence of downgrades).
+
+We ran into problems with Postgres 9.6 triggers when trying to implement an out-of-band migration to re-encode the way code intelligence data is stored. A few tables in the codeintel database stores [gob](https://golang.org/pkg/encoding/gob/)-encoded payloads in a `bytea` column. In order to migrate this data into another shape, we need to read these payloads in batches, perform the data migration over each row (in Go, which can properly decode/re-encode the data), then write the new payloads back to the database.
+
+We chose to add a `version` column to these tables so that the backend can determine how to properly interpret the payload of a particular row. This way, both old and new payload shapes can be read while new rows are written with the new shape, and old rows are _eventually_ rewritten into the new shape.
+
+This is where the tradeoffs come in to play.
+
+In our [Cloud](https://sourcegraph.com/search) environment, the `lsif_data_definitions` table contains around 300 million rows. Simply counting the number of rows takes around 24 minutes on our production database instance, as seen by the following (simplified) query plan:
+
+```
+Finalize Aggregate
+ -> Gather
+     Workers Planned: 2
+     Workers Launched: 2
+      -> Partial Aggregate
+          -> Parallel Index Only Scan using lsif_data_definitions_pkey on lsif_data_definitions
+              Heap Fetches: 14939021
+Planning Time: 1.241 ms
+Execution Time: 1442409.435 ms
+(9 rows)
+```
+
+Postgres requires that each returned row be visible to the current transaction as a way to prevent serving phantom data that was deleted by a previous query. Unfortunately, row visibility information is not stored in the index, but in the heap entry for the row.
+
+As explained in the [documentation](https://www.postgresql.org/docs/9.6/indexes-index-only-scans.html) for index-only scans, partially visible pages require a heap access to check which rows are visible and which are not. These heap access and visibility checks dominate the running time of this query.
+
+### Speeding up counting queries
+
+In no world can a nearly half-hour query be considered _efficient_. And, due to the consequences caused by Postgres [MVCC](https://www.postgresql.org/docs/9.6/mvcc-intro.html), there are few actions under our control that could make a count of so many objects faster.
```

Comment:
```suggestion
A near half-hour query is not _efficient_. Due to the consequences caused by Postgres [MVCC](https://www.postgresql.org/docs/9.6/mvcc-intro.html), there are few actions under our control that could make a count of so many objects faster. Instead, we should count fewer things.
```

Reply:
Adjusted this because we don't have access to all the perspectives in the world :)